### PR TITLE
Flare fix

### PR DESCRIPTION
**Pull request by ThislsAUsername**
I intended to make this a small PR with one fix, but you know what happens when you playtest, lol

### CODE REVIEW

File: src/Engine/MapController.java [hunk 1]
```diff
@@ -204,20 +204,22 @@ private void handleFreeTileSelect(InputHandler.InputAction input)
...
+          if( null != resident )
+          {
+            ArrayList<Unit> threats = myGame.activeCO.threatsToOverlay;
+            if( threats.contains(resident) )
+              threats.remove(resident);
+            else
+              threats.add(resident);
+          }
         }
+        myGameInputHandler.back();
```
Comment (ThatIsAUsername) [inline src/Engine/MapController.java:221]: This used to only happen `if( myGameInputHandler.isTargeting() )`. Why the change?

File: src/Engine/MapController.java [hunk 2]
```diff
@@ -383,9 +385,12 @@ public void onStateChange()
     switch (inputType)
     {
       case CONSTRAINED_TILE_SELECT:
-        // Create an option selector to keep track of where we are.
-        myGame.setCursorLocation(myGameInputHandler.getCoordinateOptions().get(myGameInputOptionSelector.getSelectionNormalized()));
+      case FREE_TILE_SELECT:
+        if( null != options && options.size() > 0 )
+          myGame.setCursorLocation(options.get(0));
```
Comment (ThatIsAUsername) [inline src/Engine/MapController.java:390]: I'm not a fan of jumping the cursor in free tile select. Also why change to getting `0` instead of `myGameInputOptionSelector.getSelectionNormalized()`?

File: src/Engine/MapController.java (hunk 1 above)
Comment (ThislsAUsername) [reply src/Engine/MapController.java:221]: I reversed that because it brought back the "unit is now very excited" after you click on it thing.
I might be able to remove it entirely in the `transaction` PR, though.

File: src/Engine/MapController.java (hunk 2 above)
Comment (ThislsAUsername) [reply src/Engine/MapController.java:390]: So, this is due to a thing I thought would be fancy: if you're in a target-dense environment (like a Rocket with a block of 3x3 units in range), you get `FREE_TILE_SELECT` instead of constrained, since the way we iterate those is a little unintuitive (and it's always going to be unintuitive for someone)

By the same token, having to move your cursor over to the enemies manually is annoying, so I fixed that too.

As for the selector: it can be null (presumably only the new situations I introduced), and it's always initialized to 0 AFAIK anyway.

File: src/Engine/GameInput/SelectActionTarget.java [hunk 3]
```diff
@@ -23,15 +23,29 @@ protected OptionSet initOptions()
...
+    boolean useFreeSelect = myStateData.actionSet.useFreeSelect;
+    // For constrained select, switch to free-tile select in target-rich environments
+    if( !useFreeSelect && targets.size() > 3 )
+    {
+      int maxDist = 0;
+      for( XYCoord a : targets )
+        for( XYCoord b : targets )
+          maxDist = Math.max(maxDist, a.getDistance(b));
+
+      useFreeSelect = maxDist < targets.size();
```
Comment (ThatIsAUsername) [inline src/Engine/GameInput/SelectActionTarget.java:36]: This is a bit weird. But OK.

File: src/Engine/MapController.java (hunk 2 above)
Comment (ThatIsAUsername) [reply src/Engine/MapController.java:390]: I get the desire to make target-dense situations a little easier to navigate, but this is no bueno for e.g. Flare and Missile Silo actions? It jumps to the NorthWest-most option, so have fun scrolling back to where you want to be.

File: src/Engine/GameInput/SelectActionTarget.java (hunk 3 above)
Comment (ThislsAUsername) [reply src/Engine/GameInput/SelectActionTarget.java:36]: It means you have to be denser than a straight line for it to "force" free tile select.
Could also make this system an option, now that we have a special page for that.

File: src/Engine/MapController.java (hunk 2 above)
Comment (ThislsAUsername) [reply src/Engine/MapController.java:390]: Bah. Good point.
Maybe this logic should go in `MapController`, then, so it can special-case constrained input as it comes in?

Comment (ThatIsAUsername) [reply src/Engine/MapController.java:390]: This _is_ MapController.

I guess you could use some kind of special-case logic though (only set the cursor for `FREE_TILE_SELECT` if there are fewer than some number of options, maybe?).

Comment (ThislsAUsername) [reply src/Engine/MapController.java:390]: To clarify:
I was meaning that I would move the "use free select if options are dense" into `MapController`, so that the rocket with friends would say to use constrained, but `MapController` would just interpret input as free select.

Actually actually: Why not (revert this change and) just make Q toggle to free select while a constrained selection is active? Best of both worlds, I think?

Comment (ThatIsAUsername) [reply src/Engine/MapController.java:390]: Maybe, except nobody would know to do that. You'd have to throw up a tooltip somewhere to say how to change modes.

File: src/Engine/MapController.java (hunk 1 above)
Comment (ThatIsAUsername) [reply src/Engine/MapController.java:221]: Does it? I don't see that.

Comment (ThislsAUsername) [reply src/Engine/MapController.java:221]: I apparently forgot to push it up. Whoops.
Bah, I need to fiddle with stuff a bit.

Comment (ThislsAUsername) [reply src/Engine/MapController.java:221]: Okay, after having doublechecked...
I think I can move it back if necessary, but we can't get rid of it entirely or you can't back out of launching flares.

File: src/Engine/MapController.java (hunk 2 above)
Comment (ThislsAUsername) [reply src/Engine/MapController.java:409]: Found a better way, I think.
It's almost like `MapController` is the place for control machinations.
